dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/datasetutility/ReplaceFileHandlerIT.java

```diff
@@ -53,6 +86,153 @@ public void shouldCreateDataFile() {
+        DataFile savedFile2 = replaceFileHandler.createDataFile(dataset, Files.readAllBytes(file2.toPath()), fileName2, fileContentType2);
+        DataFile savedFile3 = replaceFileHandler.createDataFile(dataset, Files.readAllBytes(file3.toPath()), fileName3, fileContentType3);
+
+        savedFile.setRootDataFileId(1L);
+        savedFile.setId(99L);
+        savedFile2.setRootDataFileId(1L);
+        savedFile2.setId(98L);
+        savedFile3.setRootDataFileId(1L);
+        savedFile3.setId(97L);
+
+        List<DataFile> newFiles = new ArrayList<>();
+        newFiles.add(savedFile);
+        newFiles.add(savedFile2);
+        newFiles.add(savedFile3);
+
+        fillDatasetWithExtendedData(dataset);
+        em.persist(dataset);
+
+        //when
+        ingestServiceBean.saveAndAddFilesToDataset(dataset.getEditVersion(), newFiles, dataAccess);
+
+        //then
+        Assert.assertEquals(newFiles ,dataset.getFiles());
+        Assert.assertNotSame(newFiles, dataset.getFiles());
+        Assert.assertNotEquals(new ArrayList<DataFile>(), dataset.getFiles());
+    }
+
+    @Test
+    public void shouldReplaceFile() throws IOException {
+        //given
+        dataverseSession.setUser(authenticationServiceBean.getAdminUser());
+
+        Dataverse dataverse = new Dataverse();
+        fillDataverseWithRequiredData(dataverse);
+
+        Dataset dataset = new Dataset();
+        DataAccess dataAccess = new DataAccess();
+
+        dataset.setOwner(dataverse);
+
+        File file2 = new File("/home/dataverse/dataverse4/dataverse-webapp/src/test/resources/images/banner.png");
+        String fileName2 = "banner";
+        String fileContentType2 = "png/allgood";
+
+        DataFile savedFile2 = replaceFileHandler.createDataFile(dataset, Files.readAllBytes(file2.toPath()), fileName2, fileContentType2);
+        savedFile2.setRootDataFileId(1L);
+        savedFile2.setId(99L);
+
+        List<DataFile> newFiles = new ArrayList<>();
+        newFiles.add(savedFile2);
+
+        fillDatasetWithExtendedData(dataset);
+        ingestServiceBean.saveAndAddFilesToDataset(dataset.getEditVersion(), newFiles, dataAccess);
+
+        File file3 = new File("/home/dataverse/dataverse4/dataverse-webapp/src/test/resources/images/coffeeshop.png");
+        String fileName3 = "coffeeshop";
+        String fileContentType3 = "png/allgood";
+
+        DataFile savedFile3 = replaceFileHandler.createDataFile(dataset, Files.readAllBytes(file3.toPath()), fileName3, fileContentType3);
+        newFiles.clear();
+        newFiles.add(savedFile3);
+        em.persist(dataset);
+
+        //when
+        replaceFileHandler.replaceFile(dataset.getFiles().get(0), dataset, savedFile3);
+
+        //then
+        Assert.assertEquals(2 ,dataset.getFiles().size());
+        Assert.assertEquals(savedFile3.getFileMetadatas(), dataset.getLatestVersion().getFileMetadatas());
```

Comment:
To be more sure what we are actually checking I would retrieve dataset from database and make assertions against retrieved dataset.